overlord/configstate/config/transaction.go

```diff
@@ -343,6 +402,69 @@ func commitChange(pristine *json.RawMessage, change interface{}) *json.RawMessag
 	panic(fmt.Errorf("internal error: unexpected configuration type %T", change))
 }
 
+// mergeConfigWithVirtual takes the given configuration and merges it
+// with the virtual configuration values by calling the registered
+// virtual configuration function of the given snap. The merged config
+// is returned.
+func mergeConfigWithVirtual(instanceName string, config *json.RawMessage) (*json.RawMessage, error) {
+	virtualMu.Lock()
+	km, ok := virtualMap[instanceName]
+	virtualMu.Unlock()
+	if !ok {
+		return config, nil
+	}
+
+	// create a "patch" from the virtual entries
+	patch := make(map[string]interface{})
+	for key, virtualFn := range km {
+		if virtualFn == nil {
+			continue
+		}
+		res, err := virtualFn(instanceName, key)
+		if err != nil {
+			return nil, err
+		}
+		patch[key] = jsonRaw(res)
+	}
+
+	// and apply that on top of the config
+	patchKeys := sortPatchKeysByDepth(patch)
+	for _, subkeys := range patchKeys {
+		raw := jsonRaw(patch[subkeys])
+		mergedConfig, err := PatchConfig(instanceName, strings.Split(subkeys, "."), 0, config, raw)
+		if err != nil {
+			return nil, err
+		}
+		config = jsonRaw(mergedConfig)
```

Comment:
mergedConfig should have the same as config, PatchConfig returns the same type as given for config which is already correct(?), so again maybe we don't need the full jsonRaw here